### DISCUSSION TITLE
Pass sql.DB object instead of URL

### DISCUSSION THREAD

**Discussion by gosom** [amacneil/dbmate#501] [Support]
Hello,

@amacneil and other contributors I would like to thank you for the good job.

I was looking for a way instead passing a database URL to pass an *sql.DB connection. 

This does not look possible from my quick research. 

I would like also to point out the following:

there is a [driver interface](https://github.com/amacneil/dbmate/blob/v2.8.0/pkg/dbmate/driver.go#L13) 

but when you look at the implementation of [postgres](https://github.com/amacneil/dbmate/blob/v2.8.0/pkg/driver/postgres/postgres.go#L129)

I understood that you try to connect to a default database or something . 

On top of that Open is used for every command (see Ping also for example). 

What is the reason for this behavior? 

Wouldn't be more intuitive to create a database connection and perfom the operatios with this connection ? 
This probably would allow to pass an *sql.DB already connected, which would be useful when you want to use dbmate as a library

**Comment by dossy** [comment]
Most of dbmate's functions expects a `*sql.DB` to act on.  The only ones that don't are currently `DatabaseExists()`, `CreateDatabase()`, `DropDatabase()` and `Ping()`.

It could be possible to optionally accept a `*sql.DB` parameter to those functions, in a backwards-compatible way, but as you point out, some drivers such as postgres need to open a separate connection when creating a database that doesn't exist, as you cannot connect to a database that doesn't exist yet, so we connect to the system `postgres` database to ensure we can connect, in order to create the database we need to create.

I suppose if you're invoking `CreateDatabase()` and are using dbmate as a library and want to pass an already-connected `*sql.DB`, then that means at that time the database already exists (otherwise you wouldn't be able to connect to it and wouldn't have a valid `*sql.DB` to pass in), and calling `CreateDatabase()` will be a no-op anyway, so why would you even need to invoke it?
